### ISSUE TITLE
Push Docker images into the GitHub Container Registry

### ISSUE THREAD

**Issue by ludovicm67** [matomo-org/docker#225]
I saw that this repository is configured to run GitHub Actions to build Docker images.
But for the moment it only tests if the Docker image build successfully, but they are not pushed anywhere.
It could be nice to be able to test Docker images before they are released into the official Docker images in the Docker Hub.

I suggest that we push those Docker images into the [GitHub Container Registry](https://docs.github.com/en/free-pro-team@latest/packages/getting-started-with-github-container-registry/about-github-container-registry) or in the Docker Hub in a custom organization.
I can help you to set up that if you want.

@J0WI it seems that you are in charge of the Docker images part. What do you think about that?

This could help for #221.

**Comment by mattab** [comment]
fyi @crazy-max @juliencarnot @pierreozoux @Silvenga @tianon you also have write or admin access to the repo, this may be of interest

**Comment by J0WI** [comment]
I guess this is kind of a duplicate of https://github.com/docker-library/official-images/issues/8998

**Comment by ludovicm67** [comment]
No, this is not my point. I do not ask on this issue for having the whole Docker Hub on the GitHub Container Registry.

What I asked for is to have a way to have an image very quickly, for example on each commit/tag, a CI pipeline should be triggered and an image pushed somewhere (in GitHub Container Registry or in a custom organization on the Docker Hub).

The process for publishing as an official Docker image in the Docker Hub is **really** slow. Automatic build and push of Docker image somewhere could let people try images before they are released as an official Docker image on the Docker Hub.

So to be clear, I think it could be interesting to push more regularly (for example on each commit/tag ; you can have a look [here](https://github.com/marketplace/actions/build-and-push-docker-images#handle-tags-and-labels) to see how you can configure GitHub Actions to configure tags for the Docker image) in a Docker container registry (GitHub Container Registry or a custom Docker Hub organization). I can help you on this if you need some help. This should be considered as an additional step ; this does not aim to replace the process of publishing on the Docker Hub as an official image.

**Comment by mattab** [comment]
Good to see this conversation and others. Btw just wanted to note on your offer:

>  I can help you on this if you need some help. 

That's probably very welcome! Fyi: as opposed to most other projects under matomo-org organization on Github, this docker project is entirely created and maintained by the community, and the people who build Matomo, plugins, marketplace, website, mobile app, etc. don't currently use or work on the docker project. @pierreozoux is current admin and other people are contributors:
![Screenshot from 2020-12-07 12-55-22](https://user-images.githubusercontent.com/466765/101297009-89e49280-388b-11eb-86b7-d9748b17abf4.png)

**Comment by ludovicm67** [comment]
Happy New Year!

Thanks for your answer.

If you are interested that I configure GitHub Actions to push into the GitHub Container Registry, it will require following steps:

  - Have a dedicated GitHub account in the organization and [generate a Personal Access Token](https://github.com/settings/tokens/new) for this account (just select the `write:packages` scope).
    The reason is that for the moment GitHub doesn't let us scope a personal access token to a specific organization.
    For example if I use my personal access token, the CI could publish packages here, but also on all GitHub organizations I belong + my own namespace, which is something we do not want.
    The name of the personal access token could be `matomo-org/docker` for example (in case you want to explicitly create one personal access token per project).

  - In the [Docker project](https://github.com/matomo-org/docker), go to [Settings > Secrets](https://github.com/matomo-org/docker/settings/secrets/actions).
    Create two secrets:
      - `GHCR_USERNAME` with the username of the user created in the first step,
      - `GHCR_TOKEN` with the token you got in the first step.

  - In the [packages section of the organization settings](https://github.com/organizations/matomo-org/settings/packages), enable improved container support.
    You may want to also allow members to create public packages.

  - I will work on a custom branch to write the required workflow file for GitHub Actions, and see if all is working as expected.

  - I will open a PR and you are OK with the work I did, you can merge it into `master`.

**Comment by mattab** [comment]
Happy new year 2021 everyone! 

@ludovicm67 thanks for offering your help. in general, automating some of the steps via Github actions, sounds like a good idea...

@crazy-max @juliencarnot @pierreozoux @Silvenga @tianon  maybe you have some feedback or thoughts on Ludovic's proposal above?

**Comment by Silvenga** [comment]
I wonder if matomo-org should wait for GA of the ghcr.io before using it. I have no doubt that GH will provide a better method of authenticating other then personal tokens.

In the meantime, maybe using image storage via GH Packages would work to fulfill these pain points? Those work with the github token exposed by the build pipeline (permissions are based on how the job is invoked e.g. master builds will get a token that can push containers).

https://github.com/silvenga-docker/eset-smc/blob/master/.github/workflows/build.yml#L23

My understanding is that ghcr.io will supperseed GH packages at some point - as the ghcr.io is a lot more compliant with the docker registry spec, but it is still in beta.

----

_I've been kind of out of touch with the matomo-org project lately since I switch employers._

**Comment by ludovicm67** [comment]
If you prefer, we can use a custom Docker Hub organization instead of GitHub Container Registry for the moment ; it will be easy to add it later (we can have both, or choose only one of the two ; depending on what you prefer). But we are facing the same problem as for GHCR now : we need a dedicated user on the registry for CI purposes.

If yes, we have to create a new access token on the Docker Hub (https://hub.docker.com/settings/security) using the dedicated user (first bullet point on https://github.com/matomo-org/docker/issues/225#issuecomment-753526880), and we can use `DOCKERHUB_USERNAME` and `DOCKERHUB_TOKEN` secrets instead of `GHCR_USERNAME` and `GHCR_TOKEN` (the second bullet point on https://github.com/matomo-org/docker/issues/225#issuecomment-753526880).

**Comment by ludovicm67** [comment]
Just for information, GitHub Container Registry is now generally available.

Source: https://github.blog/2021-06-21-github-packages-container-registry-generally-available/

**Comment by iamspido** [comment]
It looks like the original concern in this thread is mostly outdated now.

Back in 2020/2021, one of the practical blockers was the authentication/setup overhead around GHCR, especially the idea of needing a dedicated user plus a PAT for CI. Today, GitHub supports publishing container images to `ghcr.io` directly from GitHub Actions using `GITHUB_TOKEN` with `packages: write`, so this is no longer a major technical obstacle for the common “build on push/tag and publish from the repository workflow” setup.

So, from what I can see, the reason this still does not exist is probably not a registry limitation anymore, but simply that nobody has implemented and merged the publishing workflow yet.

Concretely, this could now be done with a fairly standard Actions setup:

* build on push/tag/release
* login to `ghcr.io` with `${{ secrets.GITHUB_TOKEN }}`
* push tags such as branch / tag / sha
* optionally publish only from default branch and version tags

Given the current state, I think this issue should be re-evaluated as an implementation/maintenance decision, not as a GitHub Container Registry limitation.

If maintainers are open to it, a PR adding GHCR publishing should be straightforward nowadays.
